### PR TITLE
322 fips 199 categorization consistency

### LINKED ISSUE #322: Check FIPS-199 Categorization Consistency with Information Types
**Extended Description**
- As an OSCAL validation producer, in order to ensure proper SSP FIPS 199 categorization has occurred, I want to have sufficient validations to accomplish that goal.

Refer to _[Guide to OSCAL-based FedRAMP System Security Plans](https://github.com/18F/fedramp-automation/blob/develop/documents/Guide_to_OSCAL-based_FedRAMP_System_Security_Plans_(SSP).pdf)_ §4.2, §4.3, §4.4.

Refer to SSP Schematron patterns `fips-140`, `sp800-60`.

The overall FIPS 199 categorization of a system is stated in SSP `<security-sensitivity-level>`.
[FIPS 199](https://csrc.nist.gov/publications/detail/fips/199/final) describes how this is determined.

The types of information maintained within a system are specified in SSP `<system-information>` `<information-type>` elements. Each `<information-type>` has base (required) and selected (optional) security impact levels.
Information types can be found in NIST [SP 800-60 v2r1](https://csrc.nist.gov/publications/detail/sp/800-60/vol-2-rev-1/final). There is also https://github.com/18F/fedramp-automation/blob/develop/dist/content/resources/xml/information-types.xml.

NIST [SP 800-60 v1r1](https://csrc.nist.gov/publications/detail/sp/800-60/vol-1-rev-1/final) describes the categorization process. Note that the process can result in adjustments. These adjustments are noted within `<information-type>` impacts.

`<security-impact-level>` is the result of the categorization process. It's subordinate elements should be consonant with `<security-sensitivity-level>` and the (composite) `<information-type>` **selected** impacts.

In other words, the categorization process resulted in selections which are consonant with one another in these SSP areas.

**Acceptance Criteria**
- [ ] Validations exist which ensure consonance within FIPS 199 and SP 800-60 SSP elements.
- [ ] All Schematron assertion messages are declarative statements which affirm the positive test outcome.
- [ ] All Schematron assertion diagnostic messages are declarative statements which explain the negative test outcome.
- [ ] The Schematron code has no assertion failures when validated using `src/validations/styleguides/sch.sch` using the `basic` phase.
- [ ] XSpec unit tests for positive and negative Schematron assertion outcomes accompany all Schematron assertions (where feasible).
- [ ] There are at least two Schematron assertions (more are quite possible depending on coding preferences) to minimally assert that
  - [ ] Every `<security-impact-level>` child element is correctly derived from the corresponding category impact set gathered from all (one or more) information types.
  - [ ] `<security-sensitivity-level>` is correctly derived from the <security-impact-level> CIA impact set.

**Story Tasks**
- [ ] Create Schematron assertions which ensure consonance within FIPS 199 and SP 800-60 SSP elements.

**Definition of Done**
 - [ ] Acceptance criteria met
 - [ ] Unit test coverage of our code > 95%  
 - [ ] Automated code quality checks passed 
 - [ ] Security reviewed and reported
 - [ ] Reviewed against plain language guidelines
 - [ ] Code must be self-documenting
 - [ ] No local tech debt
 - [ ] Load/performance tests passed – needs to be created/automated
 - [ ] Documentation updated
 - [ ] Architectural Decision Record completed as necessary for significant design choices
 - [ ] PR reviewed & approved
 - [ ] Source code merged

### PR DESCRIPTION
**Pull request by markXLIX**
closes #322

### CODE REVIEW

File: src/validations/rules/ssp.sch [hunk 1]
```diff
@@ -1616,6 +1616,62 @@
...
+                                ('')))" />
+            <sch:let
+                name="impactValue"
+                value="
+                    if ($securityImpactLevelSelected ne '')
+                    then
+                        ($securityImpactLevelSelected)
+                    else
+                        ($securityImpactLevelBase)" />
+            <sch:assert
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1668]: This assertion fails when an information-type CIA impact set is (na).

File: src/validations/rules/ssp.sch [hunk 2]
```diff
@@ -4162,11 +4202,11 @@
...
         <sch:diagnostic
-            doc:assertion="cia-impact-matches-security-sensitivity-level"
+            doc:assertion="cia-impact-matches-security-objective"
             doc:context="oscal:information-type"
-            id="cia-impact-matches-security-sensitivity-level-diagnostic">The FedRAMP SSP information type -<xsl:value-of
-                select="substring-before(local-name(), '-')" /> impact- base or selected elements do not match the security sensitivity
-            level.</sch:diagnostic>
+            id="cia-impact-matches-security-objective-diagnostic">The FedRAMP SSP security objective <xsl:value-of
+                select="substring-before(local-name(), '-')" /> does not match the <xsl:value-of
+                select="substring-before(local-name(), '-')" /> impact value.</sch:diagnostic>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:4246]: When the related assertion fails, the message `The FedRAMP SSP security objective security does not match the security impact value.` is the result.

File: src/validations/rules/ssp.sch [hunk 3]
```diff
@@ -4585,7 +4625,8 @@
         <sch:diagnostic
             doc:assertion="party-has-one-responsibility"
             doc:context="oscal:party[@type eq 'person']"
-            id="party-has-one-responsibility-diagnostic"><xsl:value-of select="o:name"/> - This person has more than one responsibility.</sch:diagnostic>
+            id="party-has-one-responsibility-diagnostic"><xsl:value-of
+                select="o:name" /> - This person has more than one responsibility.</sch:diagnostic>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:4666]: This section of code is not related to the issue. Why was it changed?

File: src/validations/rules/ssp.sch [hunk 4]
```diff
@@ -3808,7 +3843,8 @@
         <sch:diagnostic
             doc:assertion="leveraged-PE-controls-implemented-requirement"
             doc:context="/o:system-security-plan/o:control-implementation/o:implemented-requirement"
-            id="leveraged-PE-controls-implemented-requirement-diagnostic">There are PE controls inherited from leveraged authorizations.</sch:diagnostic>
+            id="leveraged-PE-controls-implemented-requirement-diagnostic">There are PE controls inherited from leveraged
+            authorizations.</sch:diagnostic>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:3884]: This section of code is not related to the issue. Why was it changed?

File: src/validations/rules/ssp.sch [hunk 5]
```diff
@@ -1537,6 +1537,33 @@
...
+                value="
+                if (matches($securityLevelStr, 'high'))
+                then
+                ('fips-199-high')
+                else
+                (if (matches($securityLevelStr, 'moderate'))
+                then
+                ('fips-199-moderate')
+                else
+                ('fips-199-low'))" />
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1554]: I see a different code formatting result when "Format and Indent" is applied to this section of code.

Are you using the settings described [here](https://github.com/18F/fedramp-automation/blob/develop/src/validations/CONTRIBUTING.md) in "Formatting XML"?

File: src/validations/rules/ssp.sch (hunk 3 above)
Comment (markXLIX) [reply src/validations/rules/ssp.sch:4666]: Change in formatting rules in oXygen.  It looks like automatic line length and space before empty tag closure.  There is no change to the code nor the output string.

File: src/validations/rules/ssp.sch (hunk 1 above)
Comment (markXLIX) [reply src/validations/rules/ssp.sch:1669]: In one of the supporting documents there is reference to any value other than the fips-199-* should be output as fips-199-low.  I cannot put my finger on it at the moment.

However, I noticed that the *-impact and security-objective-* elements both already have a test for each containing an appropriate value as defined by the fedramp_values.xml document.  So I thought that defining the else condition as 'fips-199-low' was excessive.

However, for potential clarity during maintenance I added the else condition value.

**Review by markXLIX — COMMENTED** [review]
errant click

File: src/validations/rules/ssp.sch (hunk 5 above)
Comment (markXLIX) [reply src/validations/rules/ssp.sch]: This was missed during manual formatting.  Fixed now.

File: src/validations/rules/ssp.sch (hunk 2 above)
Comment (GaryGapinski) [reply src/validations/rules/ssp.sch:4248]: This has yet to be corrected.

When the related assertion fails, the diagnostic message "The FedRAMP SSP security objective **security** does not match the **security** impact value." is the result.

"security objective security" does not seem appropriate. Neither does "security impact".

The context of the assertion is `oscal:security-objective-confidentiality | oscal:security-objective-integrity | oscal:security-objective-availability`. Perhaps you should use the third tuple of the context element names, something like `tokenize(local-name(),'-')[3]`.

Also the `doc:context` attribute value does not match the assertion context.

The following lines are currently located at 4243-4248.
```
        <sch:diagnostic
            doc:assertion="cia-impact-matches-security-objective"
            doc:context="oscal:information-type"
            id="cia-impact-matches-security-objective-diagnostic">The FedRAMP SSP security objective <xsl:value-of
                select="substring-before(local-name(), '-')" /> does not match the <xsl:value-of
                select="substring-before(local-name(), '-')" /> impact value.</sch:diagnostic>
```

Comment (markXLIX) [reply src/validations/rules/ssp.sch]: No idea what happened there as I had the change in previously.  Obviously I didn't get it into the commit.  

The correct value is now outputting.

Description: An SSP security objective value must be the same as highest available value of the same information type impact element. 
The FedRAMP SSP security objective **confidentiality** does not match the **confidentiality** impact value.

**Review by GaryGapinski — CHANGES_REQUESTED** [review]
Please correct the `cia-impact-matches-security-objective-diagnostic` diagnostic.

**Comment by markXLIX** [comment]
> Please correct the `cia-impact-matches-security-objective-diagnostic` diagnostic.

I committed changes about 10 minutes before this comment.

File: src/validations/rules/ssp.sch [hunk 6]
```diff
@@ -4244,8 +4244,8 @@
             doc:assertion="cia-impact-matches-security-objective"
             doc:context="oscal:information-type"
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:4245]: The `doc:context` remains incorrect.

**Review by GaryGapinski — CHANGES_REQUESTED** [review]

File: src/validations/rules/ssp.sch (hunk 6 above)
Comment (markXLIX) [reply src/validations/rules/ssp.sch]: fixed

**Review by GaryGapinski — APPROVED** [review]

**Pull request merged**